tests/unit/optimization/lbfgs/LBFGSMinimizer_unittest.cc

```diff
@@ -0,0 +1,285 @@
+}
+CMyCostFunction::~CMyCostFunction()
+{
+	SG_UNREF(m_obj);
+}
+void CMyCostFunction::set_target(CPiecewiseQuadraticObject *obj)
+{
+	if(obj!=m_obj)
+	{
+		SG_UNREF(m_obj);
+		m_obj=obj;
+		SG_REF(m_obj);
+	}
+}
+
+float64_t CMyCostFunction::get_cost()
+{
+	REQUIRE(m_obj,"object not set\n");
+	return m_obj->get_value();
+}
+
+CMap<TParameter*, SGVector<float64_t> >* CMyCostFunction::obtain_variable_reference()
+{
+	REQUIRE(m_obj,"object not set\n");
+	CMap<TParameter*, CSGObject*>* parameters=new CMap<TParameter*, CSGObject*>();
+	m_obj->build_gradient_parameter_dictionary(parameters);
+	index_t num_variables=parameters->get_num_elements();
+	CMap<TParameter*, SGVector<float64_t> >* variables=
+		new CMap<TParameter*, SGVector<float64_t> >(num_variables, num_variables);
+
+	for(index_t i=0; i<num_variables; i++)
+	{
+		CMapNode<TParameter*, CSGObject*>* node=parameters->get_node_ptr(i);
+		if(node->data==m_obj)
+		{
+			SGVector<float64_t> var=m_obj->get_variable(node->key);
+			variables->add(node->key,var);
+		}
+	}
+	SG_UNREF(parameters);
+	return variables;
+}
+CMap<TParameter*, SGVector<float64_t> >* CMyCostFunction::get_gradient()
+{
+	REQUIRE(m_obj,"object not set\n");
+	CMap<TParameter*, CSGObject*>* parameters=new CMap<TParameter*, CSGObject*>();
+	m_obj->build_gradient_parameter_dictionary(parameters);
+
+	index_t num_gradients=parameters->get_num_elements();
+	CMap<TParameter*, SGVector<float64_t> >* gradients=
+		new CMap<TParameter*, SGVector<float64_t> >(num_gradients, num_gradients);
+
+	for(index_t i=0; i<num_gradients; i++)
+	{
+		CMapNode<TParameter*, CSGObject*>* node=parameters->get_node_ptr(i);
+		if(node->data==m_obj)
+		{
+			SGVector<float64_t> grad=m_obj->get_gradient(node->key);
+			gradients->add(node->key,grad);
+		}
+	}
+	SG_UNREF(parameters);
+	return gradients;
+}
+
+CPiecewiseQuadraticObject::CPiecewiseQuadraticObject()
+	:CSGObject()
+{
+	init();
+}
+
+CPiecewiseQuadraticObject::~CPiecewiseQuadraticObject()
+{
+}
+
+void CPiecewiseQuadraticObject::init()
+{
+	m_init_x=SGVector<float64_t>();
+	m_truth_x=SGVector<float64_t>();
+	m_init_y=SGVector<float64_t>();
+	m_truth_y=SGVector<float64_t>();
+
+	SG_ADD(&m_init_x, "init_x", "init_x",
+		MS_AVAILABLE, GRADIENT_AVAILABLE);
+	SG_ADD(&m_init_y, "init_y", "init_y",
+		MS_AVAILABLE, GRADIENT_AVAILABLE);
+
+	SG_ADD(&m_truth_x, "truth_x", "truth_x",
+		MS_NOT_AVAILABLE);
+	SG_ADD(&m_truth_y, "truth_y", "truth_y",
+		MS_NOT_AVAILABLE);
+}
+
+void CPiecewiseQuadraticObject::set_init_x(SGVector<float64_t> init_x)
+{
+	m_init_x=init_x;
+}
+
+void CPiecewiseQuadraticObject::set_truth_x(SGVector<float64_t> truth_x)
+{
+	m_truth_x=truth_x;
+}
+
+void CPiecewiseQuadraticObject::set_init_y(SGVector<float64_t> init_y)
+{
+	m_init_y=init_y;
+}
+
+void CPiecewiseQuadraticObject::set_truth_y(SGVector<float64_t> truth_y)
+{
+	m_truth_y=truth_y;
+}
+
+float64_t CPiecewiseQuadraticObject::get_value()
+{
+	REQUIRE(m_init_x.vlen==m_truth_x.vlen, "the length must be the same\n");
```

Comment:
TOO MUCH stuff for a unit test.

Rather put these classes into Shogun code base as minimal examples how to build such cost functions, and reference them in the doxygen so that people can look things up.

But in fact @lambday what are your thoughts on making all this simpler? It is so cumbersome to define cost functions this way....